xdsl/transforms/mpi_to_llvm.py

```diff
@@ -0,0 +1,527 @@
+            tag := arith.Constant.from_int_and_width(op.tag.value.data,
+                                                     mpi.t_int),
+            comm_global :=
+            arith.Constant.from_int_and_width(self.info.mpi_comm_world_val,
+                                              mpi.t_int),
+            *(ptr := self._memref_get_llvm_ptr(op.buffer))[0],
+            func.Call.get(
+                self._mpi_name(op),
+                [ptr[1], count_ssa_val, datatype, op.dest, tag, comm_global],
+                [mpi.t_int]),
+        ], []
+
+
+class LowerMpiRecv(_MPIToLLVMRewriteBase):
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: mpi.Recv, rewriter: PatternRewriter, /):
+        rewriter.replace_matched_op(*self.lower(op))
+
+    def lower(self,
+              op: mpi.Recv) -> tuple[list[Operation], list[SSAValue | None]]:
+        """
+        This method lowers mpi.recv operations
+
+        MPI_Recv signature:
+
+        int MPI_Recv(void *buf, int count, MPI_Datatype datatype, int source, int tag,
+             MPI_Comm comm, MPI_Status *status)
+        """
+        count_ops, count_ssa_val = self._emit_memref_counts(op.buffer)
+
+        ops, new_results, status = self._emit_mpi_status_obj(
+            len(op.results) == 0)
+
+        # TODO: I really hate this dance just to make pyright happy
+        #       imo this makes code *less* readable.
+        #       The _MemRefTypeElement is bound to Attribute, so
+        #       op.buffer.typ.element_type is ALLWAYS at least Attribute!
+        assert isinstance(op.buffer.typ, memref.MemRefType)
+        memref_elm_typ = cast(memref.MemRefType[Attribute],
+                              op.buffer.typ).element_type
+
+        return [
+            *count_ops,
+            *ops,
+            *(ptr := self._memref_get_llvm_ptr(op.buffer))[0],
+            datatype := self._emit_mpi_type_load(memref_elm_typ),
+            tag := arith.Constant.from_int_and_width(op.tag.value.data,
+                                                     mpi.t_int),
+            comm_global :=
+            arith.Constant.from_int_and_width(self.info.mpi_comm_world_val,
+                                              mpi.t_int),
+            func.Call.get(self._mpi_name(op), [
+                ptr[1], count_ssa_val, datatype, op.source, tag, comm_global,
+                status
+            ], [mpi.t_int]),
+        ], new_results
+
+    # Miscellaneous
+
+
+class MpiAddExternalFuncDefs(RewritePattern):
+    mpi_func_call_names = set(_MPIToLLVMRewriteBase.MPI_SYMBOL_NAMES.values())
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, module: builtin.ModuleOp,
+                          rewriter: PatternRewriter, /):
+        # collect all func calls to MPI functions
+        funcs_to_emit: dict[str, tuple[list[Attribute],
+                                       list[Attribute]]] = dict()
+
+        @op_type_rewrite_pattern
```

Comment:
I am quite confused by this method. Isn't this just an abuse of the `match` part of the `match_and_rewrite` functionality?

IMO, this is not a great idea, as it isn't the expected usage (at least from my expectations) so it you might run into problems down the line. I personally think we should add support for ways to crawl the IRs in this manner, but not as part of the rewrite engine, but rather an analysis engine.

I am fine with leaving this in for the moment, maybe open an issue though.

Reply:
Absolutely, do you have any better solutions to this?

My two options were:
 - stateful rewrite patterns (which I somewhat disliked)
 - re-walking the whole thing (for which this was the only sane option I found)

I talked with sasha on how to do this, and we agreed that it wasn't great, but probably the best option.

Reply:
There is a `walk` function on `Operation` that you should use instead of the pattern rewriter.

Reply:
I switched to `module.walk`!